### ISSUE TITLE
Websocket compression

### ISSUE THREAD

**Issue by FZambia** [centrifugal/centrifugo#115]
Recently gorilla websocket got an update to support [compression extensions](https://tools.ietf.org/html/rfc7692) - i.e. permessage-deflate algorithm. This can reduce amount of traffic but slow down performance.

If anyone interested in this feature - please tell about it and we will add optional compression support for raw websocket. I don't need it myself at moment but someone who need can force me to do this sooner. But anyway I am going to add support for this eventually.

**Comment by FZambia** [comment]
Done in #120, it will be released with Centrifugo v1.6.0

As this compression feature considered experimental in Gorilla Websocket lib - we also should consider it experimental.

To enable raw websocket compression set `"websocket_compression": true` in config file.

**Comment by roytan883** [comment]
Can customize `websocket_compression_min_size` ? for small message no need do the compression. Currently we use 2KB as the boundary.

**Comment by FZambia** [comment]
@roytan883 as far as I understand there is an option to disable compression before each message write into connection and then turn it on again. Need to check it though

**Comment by FZambia** [comment]
@roytan883 added in 3d56eba5f9e57072931c3cad08e51f86cf7e7e18

**Comment by FZambia** [comment]
v1.6.0 released

**Comment by joshdvir** [comment]
Hi @FZambia, 

I've just installed the server and everything is running fine except for the compression, I just don't see the headers stating the permessage-deflate is working, the server is not returning them.

I've added the websocket_compression to the config file and also set websocket_compression_min_size to 1 but still no change.

Can you assist ?

Thank you.

**Comment by FZambia** [comment]
Hi! Yes, let's try to understand why it's going on.

I experimented on tornado_application from examples repo.

Here is what request headers contains:

```
Sec-WebSocket-Extensions:permessage-deflate; client_max_window_bits
```

And what Centrifugo returns in response (among other headers):

```
Sec-Websocket-Extensions:permessage-deflate; server_no_context_takeover; client_no_context_takeover
```

The first assumption - maybe you are using SockJS? Websocket compression implemented for pure websocket endpoint only.

**Comment by joshdvir** [comment]
@FZambia I'm using the javascript client with SockJS, I also see SockJS using native websockets.

Is SockJS permessage deflate not working ?

**Comment by joshdvir** [comment]
@FZambia sorry got it now, I was connected to the server SockJS endpoint, when I connected to the websocket endpoint it started working.

Thank you.

btw why is the SockJS endpoint not working with permessage deflate ?

**Comment by FZambia** [comment]
Yeah, it's not possible at moment to use websocket compression with SockJS. Because SockJS-go library does not support it yet - it requires custom to use Upgrader. I'll try to find a way to contribute ws compression into this library - this should not be difficult as it's using the same lib (Gorilla Websocket) for websockets. But at moment it's not possible unfortunately.

**Comment by joshdvir** [comment]
I might give it a go also.

Another question, I'm now POC'ing centrifugo and planning to use it on a rather big application with 200K concurrent users with billions of messages, any recommendation on your side? 

Thank you.

**Comment by FZambia** [comment]
Ah, it's a hard question, I personally never put Centrifugo into projects with such users amount. So I really looking forward for feedback from you.

You definetely need to use Redis engine and balance clients between nodes - don't forget sticky sessions for SockJS transports (needed for all except websocket), it's hard to say how many machines you'll need as it's more about message rate, channel options enabled, amount of users in channels and hardware.

At least Centrifugo is not obtrusive and can be easily swapped on another solution if you find a better. And of course we can work on bottlenecks you discover if any.

At moment largest setup I've heard was 50k users online. Maybe there are bigger numbers but I was never told about it.

**Comment by joshdvir** [comment]
Redis engine is a must and already implemented and working properly, I'm not going to use anything else except websockets for now since there is no permessage deflate and the last time I tried working without the deflate option I got around of 100TB of data transfer so it's not an option.

I'm just looking at the code of "igm/sockjs-go" and you are right it's pretty straight forward adding the deflate option to him, I might give it a go try to find the time.

I'm actually transferring from Slanger which I encounter some issues with clients subscribing to channel with him, which I'm sure won't happen here since you subscribe to 2K every time and not 100K at once which killed Redis for me, and it's hard killing Redis.


I'm going to test it in the upcoming week so I'd be happy to update on progress 😄
